### ISSUE TITLE
Error creating load balancer

### ISSUE THREAD

**Issue by font** [Azure/AKS#326]
I used Azure AKS to deploy a 3 node (default) cluster in the `centralus` location:

```bash
az aks create \
     --resource-group ${RESOURCE_GROUP} \
     --name=${CLUSTER_NAME} \
     --dns-name-prefix=${DNS_PREFIX} \
     --kubernetes-version 1.9.6 \
     --generate-ssh-keys \
     --verbose
```
Once complete, I attempted to create a LoadBalancer k8s Service, but the external IP stayed `<pending>` indefinitely. Thanks to the help from @colemickens, upon further investigation, the activity log showed it was bouncing from `Delete LoadBalancers` to `Write LoadBalancers` and both reported the error:
```
Cannot proceed with operation because resource ...ipconfig1 used by resource ...loadBalancers/kubernetes is not in Succeeded state. Resource is in Failed state and the last operation that updated/is updating the resource is PutNicOperation.
```
Investigating further using https://resources.azure.com/, showed the load balancer was failing due to the NICs on **each** of the 3 nodes. The 3 `ipconfig1` NICs on each of the nodes reported:

```yaml
"provisioningState": "Failed"
```
Attempting to kick the NICs via a GET/PUT operation on each of the NICs transitioned them to:

```yaml
"provisioningState": "Updating"
```

After a bit of waiting, they eventually went to:

```yaml
"provisioningState": "Succeeded"
```

The subsequent Activity Logs showed:

```
Write PublicIPAddresses | Succeeded | Just now | Wed Apr 25 2018 14:29:13 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | Just now | Wed Apr 25 2018 14:29:11 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | Just now | Wed Apr 25 2018 14:29:11 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | Just now | Wed Apr 25 2018 14:29:11 GMT-0700 (PDT)
Write LoadBalancers | Succeeded | Just now | Wed Apr 25 2018 14:29:11 GMT-0700 (PDT)
Delete PublicIPAddresses | Succeeded | Just now | Wed Apr 25 2018 14:29:07 GMT-0700 (PDT)
Delete LoadBalancers | Succeeded | 1 min ago | Wed Apr 25 2018 14:28:56 GMT-0700 (PDT)
Write LoadBalancers | Failed | 9 min ago | Wed Apr 25 2018 14:21:35 GMT-0700 (PDT)
Delete LoadBalancers | Failed | 10 min ago | Wed Apr 25 2018 14:20:35 GMT-0700 (PDT)
Write LoadBalancers | Failed | 14 min ago | Wed Apr 25 2018 14:16:34 GMT-0700 (PDT)
Delete LoadBalancers | Failed | 15 min ago | Wed Apr 25 2018 14:15:35 GMT-0700 (PDT)
Write LoadBalancers | Failed | 19 min ago | Wed Apr 25 2018 14:11:34 GMT-0700 (PDT)
Delete LoadBalancers | Failed | 20 min ago | Wed Apr 25 2018 14:10:35 GMT-0700 (PDT)
Write LoadBalancers | Failed | 24 min ago | Wed Apr 25 2018 14:06:34 GMT-0700 (PDT)
Delete LoadBalancers | Failed | 25 min ago | Wed Apr 25 2018 14:05:34 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | 26 min ago | Wed Apr 25 2018 14:03:50 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | 26 min ago | Wed Apr 25 2018 14:03:46 GMT-0700 (PDT)
Write NetworkInterfaces | Accepted | 27 min ago | Wed Apr 25 2018 14:02:56 GMT-0700 (PDT)
```

The NICs healed after re-submitting the NIC configs. Then the load balancer resources were succeeding, and eventually the external IP addresses showed up via `kubectl`.

I exported all the Activity log items in CSV if you're interested. Not sure what caused the NICs to originally fail to be provisioned. Any ideas?

**Comment by wilsondy** [comment]
I think we fixed it by de-allocating some public IPs.  I'm sure there's a place to find out your IP count vs quota, I just don't know where...

**Comment by frontbyte** [comment]
I've encountered a similar issue but not during a cluster creating but during creation of a service in a namespace different from default. All of them have "pending" state:
![services](https://user-images.githubusercontent.com/32958210/40118752-7277cc42-5923-11e8-8144-b2ac6253a87d.png)

Expecting the service using a command:
`kubectl --namespace=development describe service xxx-microservice`

gives me a similar error message:
`Error creating load balancer (will retry): failed to ensure load balancer for service development/xxx-microservice: timed out waiting for the condition`

**Comment by psxvoid** [comment]
I've encountered a similar issue but not during a cluster creation but during creation of a service in a namespace different from default. All of them have "pending" state:
![services](https://user-images.githubusercontent.com/32958210/40118752-7277cc42-5923-11e8-8144-b2ac6253a87d.png)

Expecting the service using a command:
`kubectl --namespace=development describe service xxx-microservice`

gives me a similar error message:
`Error creating load balancer (will retry): failed to ensure load balancer for service development/xxx-microservice: timed out waiting for the condition`

**Comment by fdc-r-wysspeter** [comment]
Same problem occurs for nginx-ingress installation:
helm install stable/nginx-ingress --name nginx-ingress --namespace kube-system --set rbac.create=false --set rbac.createRole=false --set rbac.createClusterRole=false

...then:

kubectl get service -l app=nginx-ingress --namespace kube-system

...will return <pending> for external ip.

kubectl describe service -l app=nginx-ingress --namespace kube-system

...returns:
Error creating load balancer (will retry): failed to ensure load balancer for service kube-system/nginx-ingress-controller: timed out waiting for the condition

What should I do to solve this?

**Comment by djsly** [comment]
Note that I'm not using AKS nor ACs-engine. Upstream + CloudProvider Manager

**Comment by rafaelmluz** [comment]
@peterwy01 I have the same problem. Were you able to solve the problem?

**Comment by chechiachang** [comment]
same problem here

**Comment by kblooie** [comment]
same problem here

**Comment by khsibr** [comment]
Fixed the issue by changing the SKU type of the _Public IP address_ from standard to **basic**.

**Comment by carloreggiani** [comment]
Unfortunally I have **basic** SKU, but the error is still there:

`Error creating load balancer (will retry): failed to ensure load balancer for service default/mssql-deployment: [ensure(default/mssql-deployment): lb(kubernetes) - failed to ensure host in pool: "network.InterfacesClient#CreateOrUpdate: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code=\"LinkedAuthorizationFailed\"`

**Comment by weinong** [comment]
@carloreggiani @kblooie @chechiachang @rafaelmluz can you guys provide subscriptionID, resource group, and AKS name for me to take a look? Any yaml to share with me to repro?

**Comment by shm007g** [comment]
I have this for many times when I test aks. 
Retry don't always help.

![image](https://user-images.githubusercontent.com/16999665/42752696-4caaa2ea-8921-11e8-9a15-3e3982351748.png)

**Comment by weinong** [comment]
can you guys send AKS details including subscriptionID, resourceGroup, resourceName to aks-gh-feedback@service.microsoft.com so that we can take a good look?

cc: @jdogg89 @shm007g @carloreggiani @kblooie @chechiachang @rafaelmluz

**Comment by weinong** [comment]
@rafaelmluz are you using advanced networking? k8s cloudprovider should have created the IP for you...

**Comment by rafaelmluz** [comment]
I created a traditional cluster of the k8s.... after I created public IP
and later, applied the .yaml settings.

Exactly in the step.

Em sex, 20 de jul de 2018 19:01, Weinong Wang <notifications@github.com>
escreveu:

**Comment by weinong** [comment]
hey folks, the control plane logs like controller manager logs are accessible to you for debugging

check this out!
https://medium.com/@weinong/azure-kubernetes-service-control-plane-logs-f8ffa449fd

**Comment by jdogg89** [comment]
I ended up opening a support ticket, it was an issue where when creating a
cluster with advanced networking, the service principal didn't have
contributor access in the resource group the cluster was created in. This
is obviously a bug with AKS

**Comment by yossaa** [comment]
Same issue. Caused by Public IP limit. Kubernetes reports:
Error creating load balancer (will retry): failed to ensure load balancer for service default/heat-lb: timed out waiting for the condition 

Azure activity log shows:
Cannot create more than 20 public IP addresses with static allocation method for this subscription in this region.

But... I already have one static IP -it was assigned to a service and it worked. When I changed K8S descriptor and wanted to redeploy, it stuck with this error and the old IP is still assigned.
Maybe you could change the LoadBalancer code to first remove the old IP, freeing one slot and then try to provision a new one.

**Comment by qiangli** [comment]
seems root cause is the actual creation and attaching of NIC  interface/vnet failed   on the VM node and AKS is not able to abort and recreate. Try to identify the culprit by browsing the portal.

I ended up removing the VM node and rescaling my cluster and now everything has returned to normal.

**Comment by antogh** [comment]
Hi guys I had very similar (probably the same) problem to the one reported on this thread and I solved with a simple solution, so I will describe my steps hoping to save time to others.

I created an AKS cluster for learning purposes. In my experiments I created a replica set of 3 pods and exposed it with a LB service. 1st time no problem, it worked fine, the service got its public IP and I could access the app in the pods from the public IP.
Then I deleted the service and the day after I wanted to recreate it, it was always pending failing on the public IP assignment like many of you here.

Looking at the logs in azure I saw the problem was occurring during the deletion of previous service public IP, it was not removed because still associated to the LB. Going to the LB it was marked as "Failed" and the old public IP was not listed in the frontend IP list ☹ I was feeling stuck, then I tried something very simple, I clicked refresh button on the LB, after 10 seconds … boom, it fixed it 😊, the LB was no more marked as failed and I could remove the old service public IP . Then I created the new service from kubectl normally , like 1st time, and it works fine 👍

my diagnosis of the bug is that when you delete a service of type "LoadBalancer" from kubectl , azure does not clean properly the LB frontend IP list so after that you get stuck with the old IP that cannot be removed / cannot create a new one situation. Kubectl and azure does not communicate well on lb services deletion.

**Comment by bom-sankar** [comment]
I am still getting the issue. I have deployed a 3 node cluster of 1 master and 2 worker. The I created pods for my front end, backend. All are working fine. But two of my load balancer hang on, they could not get an public Ip, below is the describe o/p

```
kubectl describe svc zevac-frontend-lb
Name:                     zevac-frontend-lb
Namespace:                default
Labels:                   <none>
Annotations:              kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"zevac-frontend-lb","namespace":"default"},"spec":{"loadBalancerIP":"52.172.46....
Selector:                 app=zevac-frontend
Type:                     LoadBalancer
IP:                       10.0.245.52
IP:                       52.172.46.210
Port:                     <unset>  80/TCP
TargetPort:               80/TCP
NodePort:                 <unset>  31723/TCP
Endpoints:                10.0.3.11:80,10.0.3.45:80
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type     Reason                      Age                From                Message
  ----     ------                      ----               ----                -------
  Warning  CreatingLoadBalancerFailed  6m (x9 over 35m)   service-controller  Error creating load balancer (will retry): failed to ensure load balancer for service default/zevac-frontend-lb: timed out waiting for the condition
  Normal   EnsuringLoadBalancer        1m (x10 over 37m)  service-controller  Ensuring load balancer
```

**Comment by selobosi** [comment]
I also spent > 1 hour solving this issue for me. API usability could be improved here. The error message 

> Error creating load balancer (will retry): failed to ensure load balancer for service &lt;service name&gt;: timed out waiting for the condition

does not provide any hint at possible causes / resolutions. If such a cause / resolution would be added, users could work around this problem more easily.

I made my way to the Activity Log in Azure Portal which gave the error code `PublicIPReferencedByMultipleIPConfigs` and an error message. With the error code and message I eventually found out that there are `Load Balancer` resources in Azure that have a `Frontend IP configuration` for each Kubernetes LoadBalancer service. These `Frontend IP configurations` had in some cases not been deleted when I deleted my service in `kubectl` and now they blocked the re-creation of my LoadBalancer services. I could solve the problem by deleting the `Frontend IP configurations` - afterwards my Kubernetes LoadBalancer services were created successfully.

**Comment by asc-adean** [comment]
I am also suffering from this problem.  The Azure load balancer does not appear to either respond to new requests after a long time waiting, and it appears to fail to delete old Frontend IP addresses long after the services were deleted in kubernetes.  I'm manually going in and comparing active service external IPs with the frontend configuration, but I should not have to do this.

**Comment by weinong** [comment]
+@feiskyer

**Comment by feiskyer** [comment]
Probably related with upstream https://github.com/kubernetes/kubernetes/issues/75198.  We have opened a PR https://github.com/kubernetes/kubernetes/pull/75256 to fix the issue (it would also be cherry picked to all stable releases).

**Comment by akin-ozer** [comment]
Probably caused from using same public ip over 2 different services, just create multiple  "az network public-ip create" and assign 1-1.

**Comment by agronholm** [comment]
> Probably caused from using same public ip over 2 different services, just create multiple "az network public-ip create" and assign 1-1.

Why is this not allowed? I found this issue looking for a solution and I have one UDP and one TCP service using the same public IP.

**Comment by akin-ozer** [comment]
Because in Kubernetes port based distribution generally made by Ingress services. You can pass that public ip to that and have many other DNS created automatically with just declaring this on service:
`loadBalancer: {}`

**Comment by allantargino** [comment]
I got the same. I my case, a tried to create an internal load balancer with a fixed IP. After while I figured it out the private IP wasn't valid in the subnet. my bad :)

**Comment by spacem** [comment]
I also just had this problem with "timed out waiting for the condition".

To resolve I created a new IP address and recreated my service using that new address so now it works. When I tried to delete the old IP address it failed and said was allocated to frontEndConfigurations so I went into the load balancer, found the front end configuration, removed all entries under "used by" then deleted the front end configuration. After this I was able to delete the old IP address.

Before this happened I did helm delete --purge, then helm install, perhaps there is some bug where delete of the service does not clear the frontEndConfiguration?

**Comment by weinong** [comment]
@spacem , what k8s version? do you have controller manager log when it was stuck?

**Comment by spacem** [comment]
@weinong it is 1.12.8 - sorry I don't believe I have additional diagnostics logging enabled

**Comment by feiskyer** [comment]
> perhaps there is some bug where delete of the service does not clear the frontEndConfiguration

There's a Kubernetes bug on https://github.com/kubernetes/kubernetes/issues/53451.

> The service controller fetches the service from the apiserver to sync. If the service no longer exists, then the load balancer is deleted. If the deletion fails (say networking issues or something), then the service key will be added back to the queue to be retried later. This works well until the service controller is restarted while a delete is in process or queued to be processed. At that point, the new service controller that takes over will never see the deleted service since the service has been deleted from etcd.

The fix https://github.com/kubernetes/kubernetes/pull/78262 would be alpha in v1.15 and beta in v1.15.

**Comment by pankajagrawal16** [comment]
What is the workaround for this issue until its fixed ? 🤔

**Comment by feiskyer** [comment]
There are a couple of issues on this page. @pankajagrawal16 Could you check kube-controller-manager logs and figure out what errors are reporting there?

**Comment by palma21** [comment]
For the number of issues in this thread:

1) 403/404 issues have been fixed in https://github.com/kubernetes/kubernetes/pull/75256 and included in v1.14.0 and above versions

2) Azure LoadBalancer Finalizer is added in https://github.com/kubernetes/kubernetes/pull/78262 as an alpha feature in v1.15.x, so it would not available until it goes to beta in next releases.

3) For Nic error issues, please open a support ticket so we can look into it.
To mitigate try to run: az network nic update


I will keep the ticket open waiting on answers to the posted questions and close afterwards. 
Please open a support ticket for your specific issues or a new issue following:
https://github.com/Azure/AKS#bug-reports-

**Comment by djsly** [comment]
thanks @palma21 , regarding 2.  will AKS support a way for the user to enable selected Alpha feature eventually?

the LB finalizer is a long waited feature and it would be a bummer if the feature stays alpha in 1.16 and we only see it available in Kubernetes 1.17, hence almost in 1 year from now for AKS support. 

If users control is out of the question for AKS, do you have run some selected Alpha feature in AKS , you Beta are the only ones allowed?

**Comment by pankajagrawal16** [comment]
@feiskyer I did not checked that before redoing everything again by destroying AKS and recreating it. That worked in next time even thought not a ideal solution. I am guess that I was too quick to purge a helm release and re deploy it again which was supposed to create a internal load balancer ip on AKS before the cleanup of load balancer was even done. 

Not sure though

**Comment by danielvdende** [comment]
We suffered a similar problem with 2 of our AKS clusters. In our case, the LoadBalancer failed. We decided to setup a new AKS cluster, to get things back up and running, while debugging in the background. We now, however, can't remove the LoadBalancer, as it's in a failed state. We asked our admins (who have sufficient access) to remove the resource group that AKS spins, but that too is now blocked thanks to this LoadBalancer that has failed... (leaving us paying for resources we can't use). I've tried several of the solutions mentioned here, but haven't found any of them to do anything unfortunately.

**Comment by palma21** [comment]
@danielvdende please open a suppor ticket as soon as possible so we can troubleshoot. Probably some operation is haging that we need to terminate, but support can help identify and act on that.


@djsly by dedinition AKS does not support alpha features, but depending on the actual maturity and stability of the feature vs. the current need/demand for it, we can make exceptions. 
We are looking into it.

**Comment by worldspawn** [comment]
Regarding Basic vs Standard this seems like it might be relevant?
https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-ip-addresses-overview-arm#sku

> Matching SKUs must be used for load balancer and public IP resources. You can't have a mixture of basic SKU resources and standard SKU resources. You can't attach standalone virtual machines, virtual machines in an availability set resource, or a virtual machine scale set resources to both SKUs simultaneously. New designs should consider using Standard SKU resources.

So I guess the azure load balancer that gets provisioned is done so with the Basic SKU?

**Comment by danielvdende** [comment]
@palma21 Yeah we've opened a support ticket, but just flagging this here, as it seems related, and it would be good for us to get some insight into how and why we've ended up in this situation (and how we can prevent it from happening again).

**Comment by palma21** [comment]
@worldspawn by default today AKS is created with BasicLB (and as such will required basic IPs).
You can create AKS using Standard LB (and standard IPs) by following: https://docs.microsoft.com/en-us/azure/aks/load-balancer-standard 

@danielvdende thank you! Will try to find it and follow up with our engineers.
This thread has forked a bit into many different (and unrelated in many cases) issues so I'd recommend opening new specific issues wo what you are experiencing if needed and if the problem is not specific to that cluster in particular.

**Comment by jluk** [comment]
Closing based on the previous comment that it is resolved. Please open a support ticket as @palma21 mentioned if you are still seeing LB creation errors and our on-call will help.

**Comment by limz8319** [comment]
> 通过将_公用IP地址_的SKU类型从标准更改为**基本**来_解决_此问题。

it works for me
